onnxscript/rewriter/basic_rules.py

```diff
@@ -123,16 +125,37 @@ def pattern(self, op, x, shape_ignored, shape):
         return op.Reshape(op.Reshape(x, shape_ignored), shape)
 
     def rewrite(self, op, x: ir.Value, shape_ignored: ir.Value, shape: ir.Value):
-        return op.Reshape(x, shape)
+        new_shape = op.initializer(ir.Tensor(self.new_shape, name=shape.name))
+        return op.Reshape(x, new_shape, allowzero=self.allowzero)
 
     def check(self, context, x, shape_ignored, shape) -> MatchResult:
         check_result = MatchResult()
-        if shape_ignored.const_value is None:
-            return check_result.fail("Shape ignored is not a constant.")
-        if shape.const_value is None:
+
+        # Shape must be a constant.
+        if (np_shape := ir_utils.get_numpy_value(shape)) is None:
             return check_result.fail("Shape is not a constant.")
-        if shape.const_value.numpy().min() <= 0:
-            return check_result.fail("Shape has non-positive values.")
+        # Convert to array to support assignment destination.
+        self.new_shape = np.array(np_shape, np_shape.dtype)
+
+        # Try to replace {0,-1} values in shape if reshape output is known.
+        if (reshape_output := context.output_values[0].shape) is not None:
+            for i, dim in enumerate(reshape_output):
+                if isinstance(dim, int) and dim > 0:
+                    self.new_shape[i] = dim
+
+        # Constraints for shape.
+        self.allowzero = context.nodes[0].attributes.get_int("allowzero", 0)
+        if self.allowzero == 1 and any(self.new_shape == 0):
+            return check_result
+        if any(self.new_shape == 0) and any(self.new_shape < 0):
+            return check_result.fail("Shape cannot contain both 0 and -1 dimensions.")
+        elif np.count_nonzero(self.new_shape == 0) > 1:
+            return check_result.fail("Shape cannot contain more than one 0 dimension.")
+
+        # At this point, we can safely replace '0' with '-1'.
+        # Note allowzero is removed since at this point it does not have any effect.
+        self.allowzero = None
```

Comment:
A context would be better. This code is not visited if the one if the above conditions is true.

Reply:
@xadupre I do not sure to understand what you mean. I'm not sure if this is clear, but I set None at this point so that when replacing the node, `allowzero` is not serialized and takes the default value (0). What do you mean by 'context would be better' ?